### PR TITLE
[gardening] Silence compiler warnings [NFC]

### PR DESCRIPTION
**Pull request by xwu**
This PR focuses on some low-hanging fruit in terms of silencing compiler warnings. Specifically:

* One use of `swap(_:_:)` is replaced with `swapAt(_:_:)` to satisfy exclusivity rules.
* Four calls to `NSDecimal{Add|Subtract|Multiply|Divide}(_:_:_:_:)` are wrapped in `withUnsafeMutablePointer(to:)` closures to satisfy exclusivity rules.
* Two unused local variables are deleted (in `Data` and `StringCore`).
* Two arithmetic expressions are rewritten so that the inferred type of integer literals is such that operations are homogeneous, avoiding warnings about deprecated `Strideable` overloads.

### CODE REVIEW

File: stdlib/public/core/Character.swift [hunk 1]
```diff
@@ -293,13 +293,13 @@ extension Character : CustomDebugStringConvertible {
 extension Character {
   @_versioned
   internal var _smallUTF16 : _UIntBuffer<UInt64, Unicode.UTF16.CodeUnit>? {
-    guard case .smallUTF16(let _63bits) = _representation  else { return nil }
+    guard case .smallUTF16(let _63bits) = _representation else { return nil }
     _onFastPath()
     let bits = UInt64(Builtin.zext_Int63_Int64(_63bits))
-    let nonZeroBitCount = type(of: bits).bitWidth - bits.leadingZeroBitCount
+    let minBitWidth = type(of: bits).bitWidth - bits.leadingZeroBitCount
```
Comment (xwu) [inline stdlib/public/core/Character.swift]: As of SE-0104, `nonzeroBitCount` is now the name for popcount, so the name of this local variable is potentially confusing; I've renamed to aid comprehension.

File: stdlib/public/SDK/Foundation/Decimal.swift [hunk 2]
```diff
@@ -205,61 +205,66 @@ extension Decimal : ExpressibleByIntegerLiteral {
 
 extension Decimal : SignedNumeric {
   public var magnitude: Decimal {
-    return Decimal(
-      _exponent: self._exponent, _length: self._length,
-       _isNegative: 0, _isCompact: self._isCompact,
-       _reserved: 0, _mantissa: self._mantissa)
+      return Decimal(_exponent: self._exponent, _length: self._length, _isNegative: 0, _isCompact: self._isCompact, _reserved: 0, _mantissa: self._mantissa)
```
Comment (CodaFi) [inline stdlib/public/SDK/Foundation/Decimal.swift]: Please (partially) restore the whitespace here.

File: stdlib/public/SDK/Foundation/Decimal.swift [hunk 3]
```diff
@@ -205,61 +205,69 @@ extension Decimal : ExpressibleByIntegerLiteral {
...
   // FIXME(integers): implement properly
   public init?<T : BinaryInteger>(exactly source: T) {
-    fatalError()
+      fatalError()
   }
 
   public static func +=(_ lhs: inout Decimal, _ rhs: Decimal) {
       var rhs = rhs
-      NSDecimalAdd(&lhs, &lhs, &rhs, .plain)
+      withUnsafeMutablePointer(to: &lhs) {
```
Comment (phausler) [inline stdlib/public/SDK/Foundation/Decimal.swift]: Thanks for taking care of these, they have been a thorn for a while now and possibly were causing some bugs.

**Review by phausler — COMMENTED** [review]
The Foundation overlay changes look good to me. The remainder should still be reviewed.

**Comment by xwu** [comment]
Sorry to ping again, @dabrahams, but would you mind reviewing the stdlib parts so that this can get merged? (I don’t have permissions to request reviewers.)

File: stdlib/public/core/Character.swift [hunk 4]
```diff
@@ -293,13 +293,13 @@ extension Character : CustomDebugStringConvertible {
...
     return _UIntBuffer<UInt64, Unicode.UTF16.CodeUnit>(
       _storage: bits,
-      _bitCount: 16 * Swift.max(1, (nonZeroBitCount + 15) / 16)
+      _bitCount: UInt8(16 * Swift.max(1, (minBitWidth + 15) / 16))
```
Comment (dabrahams) [inline stdlib/public/core/Character.swift]: If this can be done with a coercion instead of an `init()` that can trap, I'd prefer it.

**Review by dabrahams — CHANGES_REQUESTED** [review]
Looks good; just a niggle

File: stdlib/public/core/Character.swift (hunk 4 above)
Comment (xwu) [inline stdlib/public/core/Character.swift]: @dabrahams My preference too, but afaict not possible here. The problem isn’t the literals but the bit width, which is of type Int and needs to be converted and therefore may trap (unless you prefer a truncating conversion here).

Comment (dabrahams) [inline stdlib/public/core/Character.swift]: If you can prove to yourself that ordinary construction wouldn't have trapped, a truncating conversion is better.  We don't want to incur needless overheads in low-level code.

Comment (xwu) [inline stdlib/public/core/Character.swift]: You're right. Done.

**Comment by xwu** [comment]
Ready to go, I hope. Thanks all.

**Comment by xwu** [comment]
@phausler FYI, wrapping in `withUnsafeMutablePointer` results in the previously implicitly discardable result from `NSDecimal*` operations causing warnings; I have silenced that new warning with the requisite `_ =`.

**Comment by phausler** [comment]
yea we were swallowing the failures there; that perhaps was a mistake.

**Review by dabrahams — APPROVED** [review]

**Comment by xwu** [comment]
One day, maybe for Swift 5, I look forward to a holistic re-appraisal of Swift's floating point error handling mechanisms. There basically isn't one, as far as I'm aware, when it comes to `Float` and `Double`, so it's par for the course to discard these on `Decimal` as well, for now. But one day...

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by xwu** [comment]
Can I get a merge?

**Comment by dabrahams** [comment]
Thanks @xwu!

**Pull request merged**